source/opt/if_conversion.cpp

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "if_conversion.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status IfConversion::Process(ir::IRContext* c) {
+  InitializeProcessing(c);
+
+  bool modified = false;
+  std::vector<ir::Instruction*> to_kill;
+  for (auto& func : *get_module()) {
+    DominatorAnalysis* dominators =
+        context()->GetDominatorAnalysis(&func, *context()->cfg());
+    for (auto& block : func) {
+      // Get an insertion point.
+      auto iter = block.begin();
+      while (iter != block.end() && iter->opcode() == SpvOpPhi) {
+        ++iter;
+      }
+
+      ir::BasicBlock* common = nullptr;
+      block.WhileEachPhiInst([this, &modified, &common, &to_kill, dominators,
+                              &block, &iter](ir::Instruction* phi) {
+        // A false return from this function stops iteration through subsequent
+        // phis. If any of the cfg aspects of the phi are incompatible (e.g.
+        // bad domination of incoming block), there is no point checking other
+        // phi nodes. If the function returns true, then this particular phi is
+        // incompatible, but subsequent phis may be compatible.
+
+        // TODO(alan-baker): Extend to more than two predecessors
+        if (phi->NumInOperands() != 4u) return false;
+
+        // This phi is not compatible, but subsequent phis might be.
+        if (!CheckType(phi->type_id())) return true;
+
+        ir::BasicBlock* inc0 = GetIncomingBlock(phi, 0u);
+        if (dominators->Dominates(&block, inc0)) return false;
+
+        ir::BasicBlock* inc1 = GetIncomingBlock(phi, 1u);
+        if (dominators->Dominates(&block, inc1)) return false;
+
+        // All phis will have the same common dominator, so cache the result
+        // for this block. If there is no common dominator, we can stop
+        // traversing.
+        if (!common) common = CommonDominator(inc0, inc1, *dominators);
```

Comment:
This could be computed before the call to `WhileEachPhiInst` and passed in.  The CFG can give the predecessors by calling `cfg.preds(block_id)`.